substratevm/src/com.oracle.svm.agent.core/src/com/oracle/svm/agent/core/JNIHandleSetBase.java

```diff
@@ -0,0 +1,176 @@
+     * @param name Name of the method.
+     * @param signature Signature of the method. See the JNI specification for more details.
+     * @param isStatic Specifies whether the method is static or not.
+     * @return JNI method ID of the Java method.
+     */
+    protected static JNIMethodId getMethodId(JNIEnvironment env, JNIObjectHandle clazz, String name, String signature, boolean isStatic) {
+        JNIMethodId id = getMethodIdOptional(env, clazz, name, signature, isStatic);
+        guarantee(id.isNonNull());
+        return id;
+    }
+
+    /**
+     * Returns a JNI method ID of a Java method. If the method is not found, returns a
+     * {@link JNIObjectHandles#nullHandle()}.
+     *
+     * @param env JNI environment of the thread running the JVMTI callback.
+     * @param clazz Handle to the class containing the method.
+     * @param name Name of the method.
+     * @param signature Signature of the method. See the JNI specification for more details.
+     * @param isStatic Specifies whether the method is static or not.
+     * @return JNI method ID of the Java method if found, {@link JNIObjectHandles#nullHandle}
+     *         otherwise.
+     */
+    protected static JNIMethodId getMethodIdOptional(JNIEnvironment env, JNIObjectHandle clazz, String name, String signature, boolean isStatic) {
+        try (CTypeConversion.CCharPointerHolder cname = Support.toCString(name); CTypeConversion.CCharPointerHolder csignature = Support.toCString(signature)) {
+            if (isStatic) {
+                return Support.jniFunctions().getGetStaticMethodID().invoke(env, clazz, cname.get(), csignature.get());
+            } else {
+                return Support.jniFunctions().getGetMethodID().invoke(env, clazz, cname.get(), csignature.get());
+            }
+        }
+    }
+
+    /**
+     * Creates a global JNI handle of the specified local JNI handle.
+     *
+     * If the specified handle is a {@link JNIObjectHandles#nullHandle} a RuntimeException is
+     * thrown.
+     *
+     * @param env JNI environment of the thread running the JVMTI callback.
+     * @param ref A local JNI object handle.
+     * @return A global JNI object handle.
+     */
+    public JNIObjectHandle newTrackedGlobalRef(JNIEnvironment env, JNIObjectHandle ref) {
+        JNIObjectHandle global = Support.jniFunctions().getNewGlobalRef().invoke(env, ref);
+        guarantee(global.notEqual(nullHandle()));
+        globalRefsLock.lock();
+        try {
+            if (globalRefCount == globalRefs.length) {
+                globalRefs = Arrays.copyOf(globalRefs, globalRefs.length * 2);
+            }
+            globalRefs[globalRefCount] = global;
+            globalRefCount++;
+        } finally {
+            globalRefsLock.unlock();
+        }
+        return global;
+    }
+
+    /**
+     * Releases all global JNI handles. This class and the handles must not be used after a call to
+     * this method.
```

Comment:
This might be worth a `boolean destroyed` to check in asserts in the methods.

Reply:
I agree, will add it. The user should not call `destroy` manually, it should be called from the agent framework once the agent unloads. Will add a note.